Package.swift

```diff
@@ -44,11 +44,21 @@ let package = Package(
         .target(
             /** TSCBasic support library */
             name: "TSCBasic",
-            dependencies: ["TSCLibc", "TSCclibc"]),
+            dependencies: [
+                "TSCLibc",
+                "TSCclibc",
+                .product(name: "OrderedCollections", package: "swift-collections"),
```

Comment:
Just want to understand this and make sure I didn't do it wrong: Why can't I just use `"OrderedCollections"`, instead of `.product`?

SwiftPM gives this error when I try to use `"OrderedCollections"` directly:

> dependency 'OrderedCollections' in target 'TSCBasic' requires explicit declaration; reference the package in the target dependency with '.product(name: "OrderedCollections", package: "swift-collections")'

Reply:
generally, these dependencies need to be explicit (as you have). if product and package name are identical (e.g.  `.product(name: "Foo", package: "Foo")`, its enough to just use the string as short hand (e.g. `"Foo"`)

Reply:
the reason *why* it needs to be explicit is that it is designed to drive target base dependencies - and avoid needing to clone the package if its not required. one could argue the cloning is not bad but that argument can go both ways.

Reply:
cc @neonichu @abertelrud

Reply:
I believe you're right about the reason this was added, but I don't think it is just for performance — I believe that one of the motivations for the target-based dependencies is also so that it's possible to still use a package if one of its dependencies is unavailable but is not needed by products being used (typically to avoid incompatibilities that are introduced by test support libraries and other things that aren't linked into the client).  But I'm not 100% sure.

I agree, this is rather tedious, and I wonder if the situation with incompatible libraries arises often enough that it warrants the complexity.  Although I suppose that if it _does_ happen then there is no good way to resolve it absent the target-based dependency resolution.

Reply:
So I don't have actionable suggestion for this PR — those are just thoughts based on the cc: regarding this topic.

Reply:
> I believe you're right about the reason this was added, but I don't think it is just for performance — I believe that one of the motivations for the target-based dependencies is also so that it's possible to still use a package if one of its dependencies is unavailable but is not needed by products being used (typically to avoid incompatibilities that are introduced by test support libraries and other things that aren't linked into the client). But I'm not 100% sure.

what I was imagining here (and may cause other issues, so needs to be discuss in more details, probably in an amendment pitch/proposal) is that SwiftPM can _clone_ the all the dependencies and use that to build a map of product -> package before determining what packages it can then drop since they are not used. this will allow users only need to specify the product name and SwiftPM can deduce the package name on it own  (unless two packages vend the same product name)  - much like it is already doing to generate the error message!

Reply:
I just read SE-226 to learn about target-based dependency resolution. However, I'm still confused, because in SwiftPM's `Package.swift`, I can use `"OrderedCollections"` directly instead of `.product`. What's different between TSCBasic and SwiftPM's `Build` that the former needs explicit declaration for dependency while the latter doesn't?

Reply:
> much like it is already doing to generate the error message

This is also part of my original confusion. If SwiftPM can figure it out, then why do I need to specify which package a target is from.

My understanding of SE-226 is that it helps reduce the number of dependencies' dependencies to be cloned, but a target's immediate dependency packages still need to be cloned regardless (because otherwise you can't link/build the dependencies). Not specifying which package the dependency is in doesn't increase the amount of cloning, but uses more compute power to find the dependency. And it seems to me it might be a good trade-off to use a bit more compute power to reduce user friction.

Reply:
> My understanding of SE-226 is that it helps reduce the number of dependencies' dependencies to be cloned, but a target's immediate dependency packages still need to be cloned regardless (because otherwise you can't link/build the dependencies). Not specifying which package the dependency is in doesn't increase the amount of cloning, but uses more compute power to find the dependency. And it seems to me it might be a good trade-off to use a bit more compute power to reduce user friction.

personally I feel this is worth pitching as an amendment and PRing